### PR TITLE
Allow overriding tls_verify_cert_chain callback for TLS::Stream

### PR DESCRIPTION
**Pull request by hrantzsch**
Implement `set_verify_callback` in both TLS::Stream and TLS::Context.

This allows users to have the stream override the `tls_verify_cert_chain` callback. If no custom callback is set the default callback will be used.

TLS::Context is now a class and the members are protected. TLS::Stream now stores the context as a reference rather than copying it, leaving the responsibility to manage the context's lifetime with the user.

### CODE REVIEW

File: src/lib/tls/asio/asio_stream.h [hunk 1]
```diff
@@ -546,8 +567,27 @@ class Stream
...
 
+            void tls_verify_cert_chain(
+               const std::vector<X509_Certificate>& cert_chain,
+               const std::vector<std::shared_ptr<const OCSP::Response>>& ocsp_responses,
+               const std::vector<Certificate_Store*>& trusted_roots,
+               Usage_Type usage,
+               const std::string& hostname,
+               const TLS::Policy& policy) override
+               {
+               if(m_tls_context.has_verify_callback())
```
Comment (reneme) [inline src/lib/tls/asio/asio_stream.h]: Regarding my comment above: We should probably mutex this to ensure that the `m_tls_context` is not altered while performing a handshake.

File: src/lib/tls/asio/asio_context.h [hunk 2]
```diff
@@ -14,24 +14,87 @@
...
+      /**
+       * @brief Override the tls_verify_cert_chain callback
+       *
+       * This changes the verify_callback in the stream's TLS::Context, and hence the tls_verify_cert_chain callback
+       * used in the handshake.
+       * Using this function is equivalent to setting the callback via @see Botan::TLS::Stream::set_verify_callback
+       */
+      void set_verify_callback(Verify_Callback callback)
+         {
+         verifyCallback = std::move(callback);
```
[the remark this answers is not in the document]
Comment (hrantzsch) [reply src/lib/tls/asio/asio_context.h]: Yes, calling this function after initiating the async_handshake can be an issue. However, doing so is also semantically wrong, as it only makes sense to set the callback before the handshake.
I would rather add a note to point this out than mutexing in all functions involved.

File: src/lib/tls/asio/asio_context.h [hunk 3]
```diff
@@ -14,24 +14,87 @@
...
+         return static_cast<bool>(verifyCallback);
+         }
+
+   protected:
+      template <class S, class C> friend class Stream;
+
+      Credentials_Manager*   credentialsManager;
+      RandomNumberGenerator* randomNumberGenerator;
+      Session_Manager*       sessionManager;
+      Policy*                policy;
```
[the remark this answers is not in the document]
Comment (randombit) [reply src/lib/tls/asio/asio_context.h]: Yes prefer references over pointers if possible. Mostly non-owned raw pointers are used in cases when the pointer might be null, which I don't think is the case here.

[the remark this answers is not in the document]
Comment (hrantzsch) [reply src/lib/tls/asio/asio_context.h]: I agree; changed them to references.

**Comment by hrantzsch** [comment]
Noticed another issue, please don't merge yet.

**Comment by hrantzsch** [comment]
@randombit Sorry for taking a little longer on that one. We are now confident that the interface of the Context is good and won't need to be changed again all too soon. Ready for review ;)

**Comment by randombit** [comment]
Looks good!

**Pull request merged**